Fixing types, first round.

elbing opened the pull request:
Some internals headers and ap/plan9 sources

Change-Id: Ia7d02a1f81cc48ddb09ead24ca3f1625b433dfd0

elbing commented:
pdksh doesn't work yet. afree() known issue.

elbing commented:
Added new POSIX version to unistd.h

elbing commented:
New print family from musl and some types fixed more. mksh still crashes in shf_printf, it falls in a loop between internal_errorf and shf_putchar. I think because shf struct is not being filled with right flags in shf->flags (int). Dunno why.

rminnich commented on include/u.h:
```diff
@@ -14,14 +14,14 @@
-typedef	unsigned long long uvlong;
-typedef	unsigned int		Rune;
+typedef	int64_t	vlong;
+typedef	uint64_t uvlong;
```
Why define this if we don't intent do use it? Same for ulong, etc. etc.

elbing replied:
In fact, that u.h should be gone, but "out of the tree". If we use u.h we need harvey tree. It's used in ap/plan9 sources.

rminnich commented on lib/ap/internal/dir.h:
```diff
@@ -42,50 +35,50 @@ typedef	unsigned long ulong;
 void	dirtostat(struct stat *, Dir*, Fdinfo*);
-uint	convM2D(unsigned char*, uint, Dir*, char*);
-uint	convD2M(Dir*, uchar*, uint);
+uint32_t	convM2D(unsigned char*, uint32_t, Dir*, char*);
```
This is one case where specifying the big length makes no sense to me. It will fit in anything.

elbing replied:
suggestion?

rminnich commented on lib/ap/internal/dir.h:
```diff
@@ -42,50 +35,50 @@ typedef	unsigned long ulong;
-long	_dirread(int, Dir**);
-long _dirreadall(int, Dir**);
+int32_t	_dirread(int, Dir**);
+int32_t _dirreadall(int, Dir**);
```
Note of these bit length specified return values seem necessary.

rminnich commented on lib/ap/internal/fcall.h:
```diff
@@ -30,13 +30,13 @@ enum
 	char	type;
-	short	fid;
-	unsigned short	tag;
+	int16_t	fid;
```
It seems clearly to matter here, however. But fids should be unsigned. This is weird.

rminnich replied:
oh and they're 32 bits now, so I don't know why its 16 bits. here. 16 bits is old9p.

elbing replied:
Ok, uint32_t then. I'll change it :-)

rminnich commented on lib/ap/internal/fcall.h:
```diff
@@ -56,15 +56,15 @@ struct	Fcall
-			long	perm;		/* T-Create */ 
-			short	newfid;		/* T-Clone, T-Clwalk */
+			int32_t	perm;		/* T-Create */ 
+			int16_t	newfid;		/* T-Clone, T-Clwalk */
```
these fids are all wrong. What's going on here?

elbing replied:
Maybe I should compare with native ones. I just translated types as usually.

rminnich commented on lib/ap/internal/sys9.h:
```diff
@@ -91,13 +89,13 @@ enum
 extern	int	bind(const char*, const char*, int);
 #endif
-extern	int	chdir(const char*);
+extern	int	_CHDIR(const char*);
```
our chdir is compatible with linux and we should be using the same system call number.

elbing replied:
It's not for that, it's due to chdir.c is based on chdir syscall. The alternative is drop ap/plan9/chdir.c at all.

rminnich commented on lib/ap/plan9/profile.c:
```diff
@@ -25,19 +25,15 @@ enum {
-extern	void*	sbrk(unsigned long);
-extern	long	_callpc(void**);
-extern	long	_savearg(void);
+extern	void*	sbrk(uint32_t);
```
32 bits? really? makes no sense esp. for callpc.

elbing replied:
libc has uint32_t, it seemed weird to me, but...

rminnich commented on lib/ap/plan9/profile.c:
```diff
@@ -46,20 +42,20 @@ struct	Plink
 	Plink	*link;
-	long	pc;
-	long	count;
+	int32_t	pc;
```
32 bit pc?

elbing replied:
just translated. Suggestions? I think profile is not very used in APEX.

rminnich commented on lib/ap/plan9/profile.c:
```diff
@@ -46,20 +42,20 @@ struct	Plink
 {
 	void *dummy;
-	long pc;
+	int32_t pc;
```
wrong?

elbing replied:
Same as above.

rminnich commented on lib/ap/plan9/unlink.c:
```diff
@@ -42,8 +42,8 @@ unlink(const char *path)
-				sprintf(newelem, "%8.8lx%8.8lx", (ulong)(db2->qid.path>>32),
-					(unsigned long)db2->qid.path);
+				sprintf(newelem, "%8.8lx%8.8lx", (uintptr_t)(db2->qid.path>>32),
+					(uint32_t)db2->qid.path);
```
paths are 64 bits.

rminnich commented:
There's a few things here you may want to take a second look at.

elbing commented:
fprintf/vprintf from musl. Fprintf dies at vfprintf:672
testwchar dies at pwrite syscall.

removed syscall wrappers: chdir, open, close, pipe.
All tested and working.

Updated ap/internal fcall.h according to native fcall.h
(fids, etc..)

Updated ap/plan9/profile.c

elbing commented:
pending:
- sbrk -> native must be uint64_t because it's used in APEX from libc.
- dup -> uncompatible native/POSIX versions.
- wait -> uncompatible native/POSIX versions.
- fstat/stat/wstat -> untested for now.

elbing commented:
Note for myself:
- size_t must be typedef long and not long long.
- _WAIT doesn't exist!! Until we solve differences between syscall and posix implementation, ap/amd64/_wait.s must be created. wait in APEX is implemented through wait4, and this through wait harvey's syscall (ap/plan9/wait.c).

elbing commented:
Fixed size_t. dup and *stat can't be removed (better ideas without changing original syscalls?).
Remaining sbrk to 64 bit.

rminnich commented:
LGTM

I am really in awe of the amount of work that's been done here.
